### PR TITLE
Add KDE server-decoration support

### PR DESCRIPTION
**Pull request by chrisduerr**
This should add a very rough implementation of the server decoration protocol. I've tested it and it seems to work as far as I can tell. My priority was mostly to implement this for myself to force server-side decorations for GTK, this PR is more of an accident.

It's probably somewhat rough around the edges, especially since I'm not particularly familiar with all the wrappers like `Global`/`Main`/`Resource` etc, so please let me know if you have any suggestions for improvements.

### CODE REVIEW

**Comment by chrisduerr** [comment]
Another thing I forgot: I'd like to hold off on the Anvil implementation until the API is signed-off on, just to prevent doing it twice.

File: src/wayland/shell/legacy/mod.rs
Comment (i509VCB) [inline src/wayland/shell/legacy/mod.rs:67]: I think this should be `kde_decoration` to make it clear that it is kde's decoration protocol.

File: src/wayland/shell/legacy/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/legacy/decoration.rs:43]: off topic: KDE seems to have won the competition for most redundant interface names.

File: src/wayland/shell/legacy/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/legacy/decoration.rs:91]: I've noticed that when the interface is bound, the `org_kde_kwin_server_decoration_manager::default_mode` event is never sent to clients. The specificiation for the kde decorations says the following:

> This event is emitted directly after binding the interface. It contains the default mode for the decoration. When a new server decoration object is created this new object will be in the default mode until the first request_mode is requested.

File: src/wayland/shell/legacy/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/legacy/decoration.rs:77]: I think it would make more sense to have init function take the default mode as a parameter. The return value can be adjusted to also return some object that may be used to change this default mode at any time.

File: src/wayland/shell/legacy/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/legacy/decoration.rs:80]: I think this logger deserves a little use, it's trying so hard to be there!

I guess some trace or debug level messages would be fine for some debugging if needed.

File: src/wayland/shell/legacy/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/legacy/decoration.rs:99]: Wow the KDE protocol has no destructor for the global, that's going to be fun.

File: src/wayland/shell/legacy/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/legacy/decoration.rs:96]: We should probably assign the filter to the decoration manager object before we send the Setup event to users of smithay.

File: src/wayland/shell/legacy/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/legacy/decoration.rs:117]: Because the `release` request is a destructor, we can empty this match statement and instead place it in a `id.assign_destructor` or whatever it is called. See xdg-decoration, xdg-foreign and probably a couple other protocols for an example of that.

File: src/wayland/shell/legacy/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/legacy/decoration.rs:99]: There is no other request, we can place an `unreachable!()` there.

File: src/wayland/shell/legacy/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/legacy/decoration.rs:84]: Make sure you add `DispatchData<'_>` to the callback so that compositors can access their global state.

File: src/wayland/shell/legacy/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/legacy/decoration.rs:98]: Definitely a little annoying KDE's protocol does not restrict clients from binding it more than once. Also since the protocol marks it as `new_id`, we get forced to have a new object.

Comment (i509VCB) [inline src/wayland/shell/legacy/decoration.rs:98]: Because KDE's decoration protocol works in terms of `wl_surface`, I think it would be worth it to add a sort of user data storing information about a surface's current mode and updating that value as needed from the server requesting mode changes for a client asking for a different mode.

Uhh I can't recall the specifics of how we would get a user data map from the `WlSurface` at the moment, I imagine someone on matrix can figure that would or explain it further.

File: src/wayland/shell/legacy/mod.rs
Comment (chrisduerr) [reply src/wayland/shell/legacy/mod.rs:67]: Oh yeah, that's definitely one thing I wasn't sure about: Where to put it.

It's legacy, but it's not part of the legacy wl_shell. Could create a `decoration` module in the root but then I'd expect the XDG protocol to be found there too.

File: src/wayland/shell/legacy/decoration.rs
Comment (chrisduerr) [reply src/wayland/shell/legacy/decoration.rs:91]: That's a good point, I've never considered downstream just not setting this in response to the setup event. I've played around with just passing `default_mode` too but I wanted to be able to override it in the future. Both could be done, but maybe it's just easier to always set the default mode at startup and scap the Setup part, since this is the old protocol anyway?

File: src/wayland/shell/legacy/decoration.rs
Comment (chrisduerr) [reply src/wayland/shell/legacy/decoration.rs:80]: Just copied this from the XDG implementation really and thought the same. I'd assume there it at least was there to future-proof the protocol but assuming this gets merged without bugs there likely won't be any future changes that would require a logger.

Just adding some debug output might make sense, otherwise there's not much reason to have it.

File: src/wayland/shell/legacy/decoration.rs
Comment (chrisduerr) [reply src/wayland/shell/legacy/decoration.rs:96]: Why?

File: src/wayland/shell/legacy/decoration.rs
Comment (chrisduerr) [reply src/wayland/shell/legacy/decoration.rs:117]: I saw that assign_destructor call but I'm not sure what it does beyond offering a place for implementing custom destructor logic. Is it also required to free the object itself?

File: src/wayland/shell/legacy/decoration.rs
Comment (chrisduerr) [reply src/wayland/shell/legacy/decoration.rs:99]: Seems a bit counter-intuitive to defining the enums as non_exhaustive in the first place, but I guess it makes it clear to the reader that there shouldn't be any unimplemented branches.

File: src/wayland/shell/legacy/decoration.rs
Comment (chrisduerr) [reply src/wayland/shell/legacy/decoration.rs:98]: What would this be for? As a way to debounce duplicate events, potentially solve the feedback loop problem in smithay so the compositor doesn't have to worry?

File: src/wayland/shell/legacy/decoration.rs
Comment (PolyMeilex) [reply src/wayland/shell/legacy/decoration.rs:99]: > Seems a bit counter-intuitive to defining the enums as non_exhaustive in the first place, but I guess it makes it clear to the reader that there shouldn't be any unimplemented branches. Feels like you'd want a compile-time check for that stuff.

It's non_exhaustive because the protocol is unstable (and also unstandardized as a bonus)

Comment (chrisduerr) [reply src/wayland/shell/legacy/decoration.rs:99]: Yeah, I get the concept just that in this instance I'd assume things shouldn't change because XDG's protocol is the go-to for changes.

File: src/wayland/shell/legacy/mod.rs
Comment (Drakulix) [reply src/wayland/shell/legacy/mod.rs:67]: I would put this under `wayland/shell/kde_decoration`.

As long as there are client, that are actively using it, it is (sadly) not legacy. And it belongs to shells, but is not directly related to either `wl_shell` or `xdg_shell`.

File: src/wayland/shell/legacy/decoration.rs
Comment (Drakulix) [reply src/wayland/shell/legacy/decoration.rs:91]: So the "best" approach would be the most versatile:
- Provide a default-mode at `init`
- But return a handle that allows it too be changed later.

I am not sure, if this is worth for such an old protocol, but regarding @i509VCB 's comment, it should always be as hard as possible for downstream to violate/miss-use a protocol as sensibly possible. So sending the default_mode automatically at setup should be a thing.

File: src/wayland/shell/legacy/decoration.rs
Comment (Drakulix) [reply src/wayland/shell/legacy/decoration.rs:80]: Maybe log default_mode changes (if implemented)?

File: src/wayland/shell/legacy/decoration.rs
Comment (Drakulix) [reply src/wayland/shell/legacy/decoration.rs:117]: Pinging @vberger for this question.

File: src/wayland/shell/legacy/decoration.rs
Comment (Drakulix) [reply src/wayland/shell/legacy/decoration.rs:98]: Yeah basically. Also this is some kind of book-keeping, that might be useful for many compositors. So if we have the opportunity to do this for the user (and rather cheap), we would want to provide that, instead of letting downstream track the surface state manually.

File: src/wayland/shell/legacy/mod.rs
Comment (chrisduerr) [reply src/wayland/shell/legacy/mod.rs:67]: How would you feel about `shell/kde/decoration` to match the `shell/xdg/decoration`, which could also be used for other KDE-created protocols? I'm not sure if there are any other KDE protocols or if there will be in the future, but it's definitely an option.

File: src/wayland/shell/legacy/decoration.rs
Comment (chrisduerr) [reply src/wayland/shell/legacy/decoration.rs:80]: Would depend on *how* it is implemented. If downstream has direct access to the `OrgKdeKwinServerDecorationManager` then default mode can be changed without smithay ever getting any chance to log that. Though the default mode set at startup could still be logged.

File: src/wayland/shell/legacy/mod.rs
Comment (Drakulix) [reply src/wayland/shell/legacy/mod.rs:67]: Well according to [wayland.app](https://wayland.app/protocols/) there are many. I am not sure, if we want to support a lot of them, but a `kde` module will certainly make this easier later.

File: src/wayland/shell/legacy/decoration.rs
Comment (vberger) [reply src/wayland/shell/legacy/decoration.rs:117]: The main difference is that the closure given to `assign_destructor` is also invoked if the client disconnects without explicitly destroying the objects (such as when forcefully disconnected by a protocol error for example).

**Comment by chrisduerr** [comment]
I've already stated it on IRC, but I wanted to make it public here too: I currently don't plan on putting more effort into this PR, because this is basically only needed for GTK so working on the GTK patch would likely be more time-efficient.

That said, if someone else has more interest in this, feel free to clean it up. It should work as intended for anyone just trying to apply a temporary band-aid solution.

I'll leave it up to @Drakulix / @vberger if you want to keep this PR open or not. Probably going to be easier for people to find if it's still open.

**Comment by PolyMeilex** [comment]
Speaking of GTK patch, if someone is dedicated enough to work on it, here is an MR: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/2191 that can be used as a starting point (it works as is, but it needs some work to be merged). This solution would work for both GTK3 and GTK4.

I personally will not work on neither Smithay nor GTK solution, as GTK4 + libadwaita combo that will soonish become the main way of developing GTK apps, will not support SSD title bar  anyway (as libadwaita window does not have title bar by default, it is added in app source code, and therefore we don't have control over it), maybe there is some way to design nice API for libadwaita to accommodate this, but I don't feel like working on it, especially because it reached 1.0 recently. So in libadwaita app enabling SSD would basically hide shadows and that's it, I'd rather respect it's CSD preference instead.

**Comment by chrisduerr** [comment]
> I personally will not work on neither Smithay nor GTK solution, as GTK4 + libadwaita combo that will soonish become the main way of developing GTK apps, will not support SSD title bar anyway

Do you know if it still supports hiding the controls for tiled/maximized windows somehow?

**Comment by chrisduerr** [comment]
Rewrote everything from scratch to port my fork to 0.30.0. Since this forces a more standardized API, this PR might actually be mergeable now.

I don't plan to implement Anvil support, though it should be pretty trivial for someone else to do it if they're interested. But since this is basically just for legacy-support anyway I'm not sure if Anvil would want this.

File: src/wayland/shell/kde/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/kde/decoration.rs:67]: no-op is incorrect per the semantics of the protocol:

> The client can request a different mode through the decoration request. The server will acknowledge this by another event with the same mode. So even if a server prefers server-side decoration it's possible to force a client-side decoration.

File: src/wayland/shell/kde/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/kde/decoration.rs:51]: I'd prefer this to be something set in the `KdeDecorationState`. Unless it would be desirable to have per client decoration defaults.

File: src/wayland/shell/kde/decoration.rs
Comment (i509VCB) [inline src/wayland/shell/kde/decoration.rs:76]: Is it possible to provide the handle here?

**Review by i509VCB — CHANGES_REQUESTED** [review]

File: src/wayland/shell/kde/decoration.rs
Comment (chrisduerr) [reply src/wayland/shell/kde/decoration.rs:76]: What handle?

Comment (i509VCB) [reply src/wayland/shell/kde/decoration.rs:75]: The kde decoration protocol object

**Review by PolyMeilex — APPROVED** [review]
LGTM

**Review by vberger — APPROVED** [review]

**Pull request merged**
